accepted/2021/platform-dependencies/platform-dependencies.md

```diff
@@ -0,0 +1,433 @@
+
+#### Example Model
+
+```json
+{
+  "dotnetReleaseVersion": "6.0.0",
+  "platforms": [
+    {
+      "rid": "debian",
+      "components": [
+        {
+          "name": "Microsoft.NETCore.App",
+          "type": "Framework",
+          "platformDependencies": [
+            {
+              "name": "libc6",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            },
+            {
+              "name": "libgcc1",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            },
+            {
+              "name": "libgssapi-krb5-2",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            },
+            {
+              "name": "libicu57",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            },
+            {
+              "name": "liblttng-ust0",
+              "dependencyType": "LinuxPackage",
+              "usage": "diagnostics"
+            },
+            {
+              "name": "libssl1.1",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            },
+            {
+              "name": "libstdc++6",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            },
+            {
+              "name": "tzdata",
+              "dependencyType": "LinuxPackage",
+              "usage": "localization"
+            },
+            {
+              "name": "zlib1g",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            }
+          ]
+        },
+        {
+          "name": "System.DirectoryServices.Protocols",
+          "type": "NuGetPackage",
+          "platformDependencies": [
+            {
+              "name": "libldap-2.4-2",
+              "dependencyType": "LinuxPackage",
+              "usage": "default"
+            }
+          ]
+        }
+      ],
+      "platforms": [
+        {
+          "rid": "debian.10",
+          "components": [
+            {
+              "name": "Microsoft.NETCore.App",
+              "type": "Framework",
+              "platformDependencies": [
+                {
+                  "name": "libicu63",
+                  "overrides": {
+                    "name": "libicu57",
+                    "dependencyType": "LinuxPackage"
+                  }
+                }
+              ]
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+The model above uses the Debian Linux distro as an example. It shows the `Microsoft.NETCore.App` shared framework as having a base set of dependencies but for Debian 10, the `libicu57` package dependency is overriden to be `libicu63` since that is the version available in Debian 10. It also shows that the `System.DirectoryServices.Protocols` NuGet package has a dependency on `libldap-2.4-2`.
```

Comment:
libicu is tricky because we will load any version of icu found, [as long as it's at least 50](https://github.com/dotnet/runtime/blob/0e8efce08c99fb4b3ee0d46839dbe4a05c996b84/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L309). We start probing for the nonexistent libicu80 and work down. We may know what version of libicu is on each target distro, but we will load any newer one they've added - does that matter? Does that need to be expressed? I'm guessing not, since the test is - will the app run.

If there was a distro we didn't track, that had a version of libicu that was not the default on any of our target distros, we would still load it, but this json file would consider the dependency to be missing. I assume that's OK as the json is only designed to cover supported distros.

Reply:
Flipping on its head, if libicu63 is always present on Debian10, what value does it add to list it explicitly as a dependency, rather than just expressing that Debian10 will satisfy our dependencies? Will someone be constructing a docker container that has only part of Debian?

Reply:
Perhaps what I really want to know is, given supported distro XYZ, what packages do I need to explicitly add on top?

Reply:
Did you mean to mark this conversation as resolved?

> We may know what version of libicu is on each target distro, but we will load any newer one they've added - does that matter? Does that need to be expressed?

One answer to this is that we're currently [documenting](https://docs.microsoft.com/en-us/dotnet/core/install/linux-debian#dependencies) the required libicu version; we'd want to ensure that is accurate and reflects the current state of the distro's repository.  I do see there being value in having an explicit package name in general, because it obviates the need for there to be any special interpretation.  If this model is going to be used by the dotnet-install script, we don't want to have it just say "you need libicu, whatever version is fine"; you'd want it to be an explicit package name.  This makes things simple for the consuming perspective.

> Flipping on its head, if libicu63 is always present on Debian10, what value does it add to list it explicitly as a dependency, rather than just expressing that Debian10 will satisfy our dependencies? Will someone be constructing a docker container that has only part of Debian?

First, libicu is not present by default in Debian 10, at least for the Docker image.  Again, by being explicit about the dependencies you remove interpretation.  Let's say Debian 10 did include libcu.  Does the lack of including it in our model mean that we don't require or that we do but it's there by default?  That lack of information makes it difficult to do analysis of the dependencies; for example when needing to determine what the dependencies are to respond to a security incident as I've described [here](https://github.com/dotnet/designs/pull/184/files#diff-203b0b60bf87544ba5e19865ea38a42b251f78c4a904756e7499d768bb5f59d0R34).  There are other scenarios as well.  Let's say that all of the distros we support have libicu by default and so we don't describe libicu in the model as a dependency.  Now we want to evaluate another distro for inclusion in the  supported list.  Does it have the dependencies we need?  We don't know because we haven't described them fully.